### PR TITLE
Static data chip

### LINKED ISSUE #133: Static data chip
Add a chip to the basic machine which supports initializing static data into RAM. The static data chip should have a preprocessed trace containing the initial static data values. It should have an interaction with the memory chip which results in the static data being initialized into RAM.

### PR DESCRIPTION
**Pull request by morganthomas**
Closes #133

### CODE REVIEW

**Comment by morganthomas** [comment]
These changes comment out the memory argument constraints. I'm not sure this is necessary, but I found it convenient to do so since we're convinced that the memory argument is unsound anyway, and Hadas is going to replace it.

File: memory/src/lib.rs [hunk 1]
```diff
@@ -116,26 +123,44 @@ where
         // Sort first by addr, then by clk
         ops.sort_by_key(|(clk, op)| (op.get_address(), *clk));
 
-        // Consecutive sorted clock cycles for an address should differ no more
-        // than the length of the table (capped at 2^29)
-        Self::insert_dummy_reads(&mut ops);
+        // // Consecutive sorted clock cycles for an address should differ no more
+        // // than the length of the table (capped at 2^29)
+        // Self::insert_dummy_reads(&mut ops);
```
Comment (thealmarty) [inline memory/src/lib.rs:128]:
```suggestion
       // Consecutive sorted clock cycles for an address should differ no more
       // than the length of the table (capped at 2^29)
        // Self::insert_dummy_reads(&mut ops);
```
Did you mean to comment out this?

**Comment by thealmarty** [comment]
Because of the commented out code the warnings are getting out of hand. I've cleaned up as much as makes sense so we can still track them. Please merge https://github.com/valida-xyz/valida/pull/136 into this before merging this, if it's ok with you.

File: memory/src/lib.rs (hunk 1 above)
Comment (thealmarty) [reply memory/src/lib.rs:128]: I guess so otherwise there will be a warning. Feel free to ignore this.

File: memory/src/lib.rs [hunk 2]
```diff
@@ -179,6 +206,7 @@ impl MemoryChip {
 
         cols.clk = F::from_canonical_usize(clk);
         cols.counter = F::from_canonical_usize(n);
+        cols.is_static_initial = F::zero();
```
Comment (thealmarty) [inline memory/src/lib.rs:209]: I think this is not needed as it is set to zero on L204.

File: memory/src/lib.rs [hunk 3]
```diff
@@ -200,6 +228,27 @@ impl MemoryChip {
...
+    fn static_data_to_row<F: PrimeField>(
+        &self,
+        n: usize,
+        addr: u32,
+        value: Word<u8>,
+    ) -> [F; NUM_MEM_COLS] {
+        let mut row = [F::zero(); NUM_MEM_COLS];
+        let cols: &mut MemoryCols<F> = unsafe { transmute(&mut row) };
+        cols.is_static_initial = F::one();
+        cols.clk = F::zero();
```
Comment (thealmarty) [inline memory/src/lib.rs:240]: Similarly for these lines that set the values to zeros.

**Review by thealmarty — APPROVED** [review]

**Pull request merged**